main.c

```diff
@@ -489,70 +272,49 @@ asm_return_t skl_main(void)
     tpm_request_locality(tpm, 2);
     event_log_init(tpm);
 
-    /* Now that we have TPM and event log, measure bootloader data */
-    extend_pcr(tpm, &bootloader_data, bootloader_data.size, 18,
-               "Measured bootloader data into PCR18");
-
-    t = next_of_class(&bootloader_data, SKL_TAG_BOOT_CLASS);
-    if ( t == NULL || next_of_class(t, SKL_TAG_BOOT_CLASS) != NULL )
+    /*
+     * Now that we have TPM and event log, we can begin measuring. For parity
+     * with what TXT does, we leave most of measuring for the DLME. Here, we
+     * only have to measure DLME itself, as well as entry point offset - only
+     * both of those measurements together tell that the proper code has been
+     * started. Entry point offset may come from MLE header included in DLME,
+     * but we can't trust that the bootloader passed it without modification.
+     */
+    dl_info = next_entry_with_tag(NULL, SLR_ENTRY_DL_INFO);
+
+    if ( dl_info                                     == NULL
+         || dl_info->hdr.size                        != sizeof(*dl_info)
+         || end_of_slrt()                             < _p(&dl_info[1])
+         || dl_info->dlme_base                       >= 0x100000000ULL
+         || dl_info->dlme_base + dl_info->dlme_size  >= 0x100000000ULL
+         || dl_info->dlme_entry                      >= dl_info->dlme_size )
     {
-        print("No boot tag or multiple boot tags\n");
+        print("Bad bootloader data format\n");
         reboot();
     }
 
-    switch( t->type )
-    {
-    case SKL_TAG_BOOT_LINUX:
-        ret = skl_linux(tpm, (struct skl_tag_boot_linux *)t);
-        break;
-    case SKL_TAG_BOOT_MB2:
-        ret = skl_multiboot2(tpm, (struct skl_tag_boot_mb2 *)t);
-        break;
-    case SKL_TAG_BOOT_SIMPLE:
-        ret = skl_simple_payload(tpm, (struct skl_tag_boot_simple_payload *)t);
-        break;
-    default:
-        print("Unknown kernel boot protocol\n");
-        reboot();
-    }
+    entry_offset = dl_info->dlme_entry;
+    extend_pcr(tpm, &entry_offset, sizeof(entry_offset), 17,
+               "DLME entry offset");
+    extend_pcr(tpm, _p(dl_info->dlme_base), dl_info->dlme_size, 17, "DLME");
 
     tpm_relinquish_locality(tpm);
     free_tpm(tpm);
 
+    ret.dlme_entry = _p(dl_info->dlme_base + dl_info->dlme_entry);
+    ret.dlme_arg = _p(dl_info->bl_context.context);
```

Comment:
Maybe worth checking that bootloader ID is `SLR_BOOTLOADER_GRUB` (`1`) given that this field is bootloader-specific.

Reply:
Added a bit higher: https://github.com/TrenchBoot/secure-kernel-loader/compare/f2a18cb10d6548d177ca1eff9b8334344b2a306b..b9aaa5c37b91adeed72a067dde349129b9970fde#diff-a0cb465674c1b01a07d361f25a0ef2b0214b7dfe9412b7777f89add956da10ecR291